src/Umbraco.Web.Common/ActionsResults/PublishedContentNotFoundResult.cs

```diff
@@ -0,0 +1,60 @@
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Umbraco.Web.Routing;
+
+namespace Umbraco.Web.Common.ActionsResults
+{
+    /// <summary>
+    /// Returns the Umbraco not found result
+    /// </summary>
+    public class PublishedContentNotFoundResult : IActionResult
+    {
+        private readonly IUmbracoContext _umbracoContext;
+        private readonly string _message;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PublishedContentNotFoundResult"/> class.
+        /// </summary>
+        public PublishedContentNotFoundResult(IUmbracoContext umbracoContext, string message = null)
+        {
+            _umbracoContext = umbracoContext;
+            _message = message;
+        }
+
+        /// <inheritdoc/>
+        public async Task ExecuteResultAsync(ActionContext context)
+        {
+            HttpResponse response = context.HttpContext.Response;
+
+            response.Clear();
+
+            response.StatusCode = StatusCodes.Status404NotFound;
+
+            IPublishedRequest frequest = _umbracoContext.PublishedRequest;
+            var reason = "Cannot render the page at URL '{0}'.";
+            if (frequest.HasPublishedContent() == false)
+            {
+                reason = "No umbraco document matches the URL '{0}'.";
+            }
+            else if (frequest.HasTemplate() == false)
+            {
+                reason = "No template exists to render the document at URL '{0}'.";
+            }
+
+            await response.WriteAsync("<html><body><h1>Page not found</h1>");
+            await response.WriteAsync("<h2>");
+            await response.WriteAsync(string.Format(reason, WebUtility.HtmlEncode(_umbracoContext.OriginalRequestUrl.PathAndQuery)));
+            await response.WriteAsync("</h2>");
+            if (string.IsNullOrWhiteSpace(_message) == false)
+            {
+                await response.WriteAsync("<p>" + _message + "</p>");
+            }
+
+            await response.WriteAsync("<p>This page can be replaced with a custom 404. Check the documentation for \"custom 404\".</p>");
```

Comment:
I know it is not a new thing, but we should make a umbra.co/404-docs link and point to the docs

Reply:
fixed

Reply:
I changed to a tiny link, so we can change the redirect without updating umbraco if we someday need that..